pkgs/stdenv/linux/default.nix

```diff
@@ -284,21 +282,63 @@ in
     };
   })
 
-
   # 2nd stdenv that contains our own rebuilt binutils and is used for
-  # compiling our own Glibc.
+  # compiling our own gcc.
   #
   # resulting stage2 stdenv:
-  # - coreutils, glibc, gcc: from bootstrapFiles
-  # - binutils: from nixpkgs, built by bootstrapFiles toolchain
+  # - coreutils, glibc: from bootstrapFiles
+  # - binutils, gcc: from nixpkgs, built by bootstrapFiles toolchain
   (prevStage: stageFun prevStage {
     name = "bootstrap-stage2";
 
     overrides = self: super: {
       inherit (prevStage)
         ccWrapperStdenv
-        gcc-unwrapped coreutils gnugrep
-        perl gnum4 bison;
+        bash bison gnum4 coreutils gnugrep
+        patchelf perl texinfo xz;
+
+      ${localSystem.libc} = getLibc prevStage;
+
+      gcc-unwrapped =
+        let makeStaticLibrariesAndMark = pkg:
+              lib.makeOverridable (pkg.override { stdenv = self.makeStaticLibraries self.stdenv; })
+                .overrideAttrs (a: { pname = "${a.pname}-stage2"; });
```

Comment:
I'm seeing a whole lot of packages being rebuilt an extra time with this PR -- `binutils`, `gettext`, and others.

Please mark the extra builds with `-stageX` in their `pname` like above so we can tell them apart, and so we can understand the rebuild impact of this PR.  Right now we don't have a good handle on it.

For example:

```shell
# before this PR
$ nix-store -q --requisites $(nix-instantiate . -A stdenv --argstr system powerpc64le-linux) | grep binutils-2.39.drv | wc -l
3
# with this PR
$ nix-store -q --requisites $(nix-instantiate . -A stdenv --argstr system powerpc64le-linux) | grep binutils-2.39.drv | wc -l
4
```

Reply:
Helpful shell command:

```
nix-store -q --requisites $(nix-instantiate . -A stdenv) \
| cut -b45- \
| sort \
| grep '\.drv$' \
| uniq -D
```

Based on a diff f the output of the command above before and after, it looks like we've added:

- *two* more rebuilds of `gcc` (one expected)
- one more rebuild of `binutils`
- one more rebuild of `gettext` (ugh)
- two more rebuilds of `gmp-with-cxx`
- two more rebuilds of `mpfr`
- one more build of `libxcrypt`
- one more build of `zlib`
- two more rebuilds of `which`

Reply:
> Using uniq -c here is misleading.

No, `uniq -c` removes identical duplicate store paths when target of an arrow is the same store path with multiple inputs. Store hash is removed after the `uniq`. It should tell the correct numbers of rebuilds (maybe slightly outdated as I added extra `xz` rebuild since).

> two more rebuilds of gcc (only one expected -- update the PR title?)

No, I still think it introduces one extra `gcc` build. Building it twice is not very trivial from standpoint of used references. You can check which exact stdenv triggers the builds to make sure:

```
$ nix-store --query --graph $(nix-instantiate -A stdenv) |& grep -P 'stdenv.* -> .*gcc-11'

"3bxxwn8hbz725i081fqlylqljyp5pk8m-bootstrap-stage4-stdenv-linux.drv" -> "1ikpv8jbf5z18hp0gkiw27234ifvgmbk-gcc-11.3.0.drv" [color = "red"];
"xck2dhlvi35c0lnc762z57sh1lv22n2d-bootstrap-stage2-stdenv-linux.drv" -> "imhrq50yi7mkx4wgbv6sifjyhbi1yymn-gcc-11.3.0.drv" [color = "red"];
```

Tl;DR: one in stage2 (new addition) and one in stage4 (pre-existing one).

Or you can look up `gcc` nodes in slightly filtered `.svg` file to ease visual inspection:

```
$ nix-store --query --graph $(nix-instantiate -A stdenv) |
    grep -vP '[.](sh|tar|bash|patch|c|diff)|bash52-0|wrapper|hook|expand-response-params' |
    dot -Tsvg > /tmp/a.svg; firefox /tmp/a.svg
```

I still claim we have only 2 `gcc` builds here: https://trofi.github.io/posts.data/275-nixpkgs-bootstrap-deep-dive/08-nixpkgs-bootstrap-tree.svg.